File: Data Structures & Algorithms/subsets/submission-1.cpp
class Solution {
public:
    vector <vector<int>> res;
    vector<int> curr;

    void backtrack(int idx, vector<int>& nums){
        if(idx == nums.size()){
            res.push_back(curr);
            return;
        }
        curr.push_back(nums[idx]);
        backtrack(idx+1, nums);
        curr.pop_back();
        backtrack(idx+1, nums);
        

    }
    vector<vector<int>> subsets(vector<int>& nums) {
        if(nums.size() == 0) return res;
        backtrack(0,nums);
        return res;

        
    }
};

File: Data Structures & Algorithms/combinations-of-a-phone-number/submission-1.java
class Solution {
    List<String> res = new ArrayList<>();
    StringBuilder curr = new StringBuilder() ;
    List<String> pad = List.of(
    "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"
);
    void backtrack(String s, int idx){
        if (idx == s.length()){
            res.add(curr.toString());
            return;
        }
        int tmp = s.charAt(idx) - '0';
        for (char p : pad.get(tmp).toCharArray()){
            curr.append(p);
            backtrack(s, idx+1);
            curr.deleteCharAt(curr.length()-1);
        }
    }
    public List<String> letterCombinations(String digits) {
       if(digits.length() == 0) return res;
        backtrack(digits, 0);
        return res;
    }
}
